### ISSUE TITLE
CSS media queries are not detected (thrown away) if there are CSS comments outside CSS selector blocks

### ISSUE THREAD

**Issue by NeoEGM** [jjriv/emogrifier#154]
# Example code

**Try to emogrify the following responsive HTML:**
https://github.com/mailchimp/email-blueprints/blob/master/responsive-templates/base_boxed_basic_query.html

# Where is the problem

**On the private member function "splitCssAndMediaQuery".**
The regular expression doesn't account for comments outside selector blocks. Specially after them, before the final closing "}" for the media query block.
It **then** removes the comments. After the regex matching. So media queries with this syntax are lost.

```
private function splitCssAndMediaQuery($css)
{
	$media = '';

	$css = preg_replace_callback(
		'#@media\\s+(?:only\\s)?(?:[\\s{\\(]|screen|all)\\s?[^{]+{.*}\\s*}\\s*#misU',
		function($matches) use (&$media) {
			$media .= $matches[0];
		},
		$css
	);

	// filter the CSS
	$search = array(
		// get rid of css comment code
		'/\\/\\*.*\\*\\//sU',
		// strip out any import directives
		'/^\\s*@import\\s[^;]+;/misU',
		// strip remains media enclosures
		'/^\\s*@media\\s[^{]+{(.*)}\\s*}\\s/misU',
	);

	$replace = array(
		'',
		'',
		'',
	);

	// clean CSS before output
	$css = preg_replace($search, $replace, $css);

	return array('css' => $css, 'media' => $media);
}
```

# Fix

**Remove the comments *before* the regex matching.**

```
private function splitCssAndMediaQuery($css)
{
	$media = '';

	//Remove comments previously to matching media queries
	$css = preg_replace('/\\/\\*.*\\*\\//sU', '', $css);

	$css = preg_replace_callback(
		'#@media\\s+(?:only\\s)?(?:[\\s{\\(]|screen|all)\\s?[^{]+{.*}\\s*}\\s*#misU',
		function($matches) use (&$media) {
			$media .= $matches[0];
		},
		$css
	);

	// filter the CSS
	$search = array(
		// strip out any import directives
		'/^\\s*@import\\s[^;]+;/misU',
		// strip remains media enclosures
		'/^\\s*@media\\s[^{]+{(.*)}\\s*}\\s/misU',
	);

	$replace = array(
		'',
		'',
	);

	// clean CSS before output
	$css = preg_replace($search, $replace, $css);

	return array('css' => $css, 'media' => $media);
}
```

# Workaround

**Remove CSS comments from HTML *before* passing them to Emogrifier.**

## I've written a function to do this:

```
function RemoveCSSComments($html)
{
	try
	{
		$dom = new DOMDocument();
		
		$dom->preserveWhitespace = false;
		$dom->formatOutput = true;
		
		if (!@$dom->loadHTML($html))
			return false;
		
		//Search every <style> node
		$xpath = new DOMXPath($dom);
		$styleNodes = $xpath->query('//style');

		if ($styleNodes !== false)
			foreach ($styleNodes as $styleNode)
				$styleNode->nodeValue = RemoveCommentsFromCSSCode($styleNode->nodeValue);
		
		return $dom->saveHTML();
	}
	catch (DOMException $e)
	{
		return false;
	}
}

function RemoveCommentsFromCSSCode($css)
{
	return preg_replace('/\\/\\*.*\\*\\//sU', '', $css);
}
```

## You just call it:

`$html = RemoveCSSComments($html);`

**Comment by oliverklee** [comment]
Hi NeoEGM, thanks a lot for the detailed bug report! I'll have a look at it.

**Comment by Renkas** [comment]
Why convert input to DOMDocument, then back to string so that it could be converted to DOMDocument again? :)

Wouldn't it be better to just run RemoveCommentsFromCSSCode() on the whole input HTML string?

**Comment by oliverklee** [comment]
Yes, of course. Thanks for noticing!

@Renkas Would you be interested in putting the concept into a PR?

**Comment by Renkas** [comment]
Sorry - that was the workaround not the fix that did double converting. I added the fix as suggested by @NeoEGM and also added a testcase.

**Comment by oliverklee** [comment]
Hi @NeoEGM, I've polished and merged the fix for this bug in #201. Thanks for your contribution of reporting the bug and proposing the solution!

**Comment by NeoEGM** [comment]
You're welcome!
